### ISSUE TITLE
Implement Imitation Learning Pre-training + RL Fine-tuning

### ISSUE THREAD

**Issue by BY571** [TorchTrade/torchtrade#54]
## Problem

PPO agents start from **random initialization**, wasting thousands of steps learning basic trading concepts ("don't always hold", "take profits", "cut losses").

**Evidence:**
- First 10k-20k frames spent learning obvious strategies
- Random exploration rarely discovers profitable patterns early
- No prior knowledge of what "good trading" looks like

---

## Proposed Solution: Imitation Learning + RL

**Two-stage training:**
1. **Pre-training**: Learn from expert demonstrations (imitation learning)
2. **Fine-tuning**: Improve with RL (PPO) using environment rewards

### Why This Works

**Problem with pure RL:**
```
Random agent → Explore randomly → Eventually find good strategy → Optimize
                ^^^^^^^^^^^^^^^^^
                10k-20k frames wasted
```

**With imitation pre-training:**
```
Expert demos → Imitate good strategies → Start from reasonable baseline → RL fine-tune
               ^^^^^^^^^^^^^^^^^^^^^^^^
               Bootstrap with prior knowledge
```

---

### Option A: Rule-Based Experts

Create simple, profitable heuristics as "experts":

```python
class MomentumExpert:
    """Simple momentum trading strategy."""
    
    def select_action(self, observation):
        # Extract features
        prices = observation["market_data"][:, -1]  # Close prices
        returns = torch.diff(prices) / prices[:-1]
        
        # Calculate momentum
        momentum = returns[-10:].mean()  # 10-bar momentum
        volatility = returns[-20:].std()  # 20-bar volatility
        
        # Decision logic
        if momentum > 0.01 and volatility < 0.02:
            return 2  # Long moderate (Action 2)
        elif momentum < -0.01 and volatility < 0.02:
            return 5  # Short moderate (Action 5)
        else:
            return 0  # Hold

class MeanReversionExpert:
    """Mean reversion strategy."""
    
    def select_action(self, observation):
        prices = observation["market_data"][:, -1]
        current_price = prices[-1]
        ma_20 = prices[-20:].mean()
        
        # Calculate deviation from moving average
        deviation = (current_price - ma_20) / ma_20
        
        if deviation < -0.02:  # Oversold
            return 1  # Long conservative (Action 1)
        elif deviation > 0.02:  # Overbought
            return 4  # Short conservative (Action 4)
        else:
            return 0  # Hold

class BreakoutExpert:
    """Breakout/volatility expansion strategy."""
    
    def select_action(self, observation):
        prices = observation["market_data"][:, -1]
        highs = observation["market_data"][:, 1]
        lows = observation["market_data"][:, 2]
        
        # Bollinger Bands
        ma = prices[-20:].mean()
        std = prices[-20:].std()
        upper_band = ma + 2 * std
        lower_band = ma - 2 * std
        
        current_price = prices[-1]
        
        if current_price > upper_band:
            return 3  # Long aggressive (Action 3)
        elif current_price < lower_band:
            return 6  # Short aggressive (Action 6)
        else:
            return 0  # Hold
```

### Collect Demonstrations

```python
def collect_expert_demonstrations(
    env,
    experts: List[Callable],
    num_episodes: int = 1000,
) -> TensorDict:
    """Collect state-action pairs from expert policies."""
    
    demonstrations = []
    
    for expert in experts:
        for _ in range(num_episodes // len(experts)):
            obs = env.reset()
            done = False
            
            while not done:
                # Expert selects action
                action = expert.select_action(obs)
                
                # Store demonstration
                demonstrations.append({
                    "observation": obs.clone(),
                    "action": action,
                    "expert_id": experts.index(expert),
                })
                
                # Step environment
                obs = env.step(TensorDict({"action": action}))
                done = obs["done"]
    
    return TensorDict.stack(demonstrations)
```

### Option B: Human Demonstrations (Advanced)

- Record human trading decisions via UI
- Replay historical successful trades
- Use labeling service (humans label actions for given states)

---

## Phase 2: Imitation Learning (Behavioral Cloning)

Train policy to mimic expert actions:

```python
class BehavioralCloning:
    """Train policy via supervised learning on expert demos."""
    
    def __init__(self, policy_network, optimizer):
        self.policy = policy_network
        self.optimizer = optimizer
        self.criterion = nn.CrossEntropyLoss()
    
    def train(
        self,
        demonstrations: TensorDict,
        epochs: int = 50,
        batch_size: int = 256,
    ):
        """Train policy to predict expert actions."""
        
        dataloader = DataLoader(demonstrations, batch_size=batch_size, shuffle=True)
        
        for epoch in range(epochs):
            total_loss = 0
            total_accuracy = 0
            
            for batch in dataloader:
                obs = batch["observation"]
                expert_actions = batch["action"]
                
                # Forward pass
                logits = self.policy(obs)
                loss = self.criterion(logits, expert_actions)
                
                # Backward pass
                self.optimizer.zero_grad()
                loss.backward()
                self.optimizer.step()
                
                # Metrics
                pred_actions = logits.argmax(dim=-1)
                accuracy = (pred_actions == expert_actions).float().mean()
                
                total_loss += loss.item()
                total_accuracy += accuracy.item()
            
            print(f"Epoch {epoch}: Loss={total_loss:.4f}, Acc={total_accuracy:.4f}")
```

---

## Phase 3: RL Fine-tuning

Continue training with PPO, starting from BC-initialized policy:

```python
def train_with_imitation_pretraining(
    env,
    experts: List[Callable],
    config,
):
    # Phase 1: Collect demonstrations
    print("Phase 1: Collecting expert demonstrations...")
    demonstrations = collect_expert_demonstrations(env, experts, num_episodes=1000)
    
    # Phase 2: Behavioral cloning pre-training
    print("Phase 2: Pre-training with behavioral cloning...")
    policy, critic = create_ppo_models(config)
    bc = BehavioralCloning(policy, optimizer=torch.optim.Adam(policy.parameters(), lr=1e-3))
    bc.train(demonstrations, epochs=50)
    
    # Evaluate pre-trained policy
    print("Evaluating BC policy before RL...")
    bc_performance = evaluate(env, policy)
    print(f"BC Policy Sharpe: {bc_performance:.2f}")
    
    # Phase 3: RL fine-tuning
    print("Phase 3: Fine-tuning with PPO...")
    train_ppo(env, policy, critic, config, total_frames=100_000)
    
    # Evaluate final policy
    final_performance = evaluate(env, policy)
    print(f"Final Policy Sharpe: {final_performance:.2f}")
```

---

### Learning Curve Comparison

```
Random Init PPO:
  Frames 0-10k:    Sharpe ~ -10 (random flailing)
  Frames 10k-40k:  Sharpe ~ -5  (learning basics)
  Frames 40k-100k: Sharpe ~ -2  (optimizing)

BC + PPO:
  Frames 0:        Sharpe ~ -2  (BC initialized)
  Frames 0-20k:    Sharpe ~ -1  (quick improvement)
  Frames 20k-100k: Sharpe ~ 0   (fine-tuning)
```

### Success Metrics

| Metric | Random Init | BC + PPO Target |
|--------|------------|-----------------|
| Initial policy Sharpe | -10 to -15 | -2 to 0 |
| Frames to Sharpe > -2 | 40k-60k | 0 (immediate) |
| Final Sharpe (100k frames) | -2 to -5 | > 0 |
| Action diversity (early) | Low | High (experts use diverse actions) |

---

## Advantages

**vs Pure RL:**
- ✅ **Faster learning**: Start from reasonable baseline
- ✅ **Better exploration**: Experts demonstrate diverse strategies
- ✅ **More sample efficient**: Less random flailing
- ✅ **Interpretable**: Can analyze what experts teach

**vs Pure Imitation:**
- ✅ **Adapts beyond experts**: RL can improve on heuristics
- ✅ **Learns from environment**: Not limited to expert knowledge
- ✅ **Optimizes for true objective**: Environment rewards, not demo matching

---

### Week 1: Expert Strategies
- [ ] Implement 3 rule-based experts (Momentum, MeanReversion, Breakout)
- [ ] Test experts on historical data
- [ ] Collect 1000 episodes of demonstrations per expert
- [ ] Analyze expert action distributions and performance

### Week 2: Behavioral Cloning
- [ ] Implement BC training loop
- [ ] Train policy on expert demos (50 epochs)
- [ ] Evaluate BC policy on test set
- [ ] Compare BC vs random init baseline

### Week 3: RL Fine-tuning
- [ ] Fine-tune BC policy with PPO (100k frames)
- [ ] Compare learning curves: BC+PPO vs random init PPO
- [ ] Analyze what RL learned beyond BC
- [ ] Test on different market regimes

---

## Configuration Example

```yaml
# config.yaml
imitation:
  enabled: true
  
  # Expert demonstration collection
  experts: ["momentum", "mean_reversion", "breakout"]
  demo_episodes_per_expert: 333  # Total 1000 episodes
  
  # Behavioral cloning pre-training
  bc_epochs: 50
  bc_batch_size: 256
  bc_learning_rate: 1e-3
  
  # RL fine-tuning
  finetune_frames: 100000
  finetune_from_checkpoint: "bc_policy.pt"
```

---

### Momentum Expert Performance
```
Expected Sharpe: 0.5 to 1.0 (simple but works in trends)
Action distribution: 40% long, 40% short, 20% hold
Risk: Underperforms in ranging markets
```

### Mean Reversion Expert Performance
```
Expected Sharpe: 0.3 to 0.8 (works in ranges)
Action distribution: 30% long, 30% short, 40% hold
Risk: Fails in strong trends
```

### Breakout Expert Performance
```
Expected Sharpe: 0.2 to 1.5 (high variance, captures big moves)
Action distribution: 25% long aggressive, 25% short aggressive, 50% hold
Risk: Many false breakouts
```

**Combined ensemble Sharpe: ~1.0-1.5** (better than any single expert)

---

### DAgger (Dataset Aggregation)

Iteratively improve demonstrations:

```python
# Round 1: Train on expert demos
# Round 2: Run policy, collect states where it diverges from expert
# Round 3: Ask expert to label those states
# Round 4: Retrain on expanded dataset
# Repeat
```

### GAIL (Generative Adversarial Imitation Learning)

Learn reward function from demos:

```python
# Discriminator learns to distinguish expert vs policy actions
# Policy optimized to fool discriminator
# More robust than BC (doesn't require perfect expert labeling)
```

---

## Monitoring & Debugging

Track these metrics:

```python
metrics = {
    # BC phase
    "bc/loss": bc_loss,
    "bc/accuracy": bc_accuracy,
    "bc/eval_sharpe": bc_sharpe,
    
    # RL phase
    "finetune/sharpe": ppo_sharpe,
    "finetune/action_divergence": kl_div(bc_actions, ppo_actions),
    "finetune/improvement": ppo_sharpe - bc_sharpe,
}
```

---

## References

**Imitation Learning:**
- Pomerleau (1988) - "ALVINN: An Autonomous Land Vehicle in a Neural Network"
- Ross et al. (2011) - "A Reduction of Imitation Learning and Structured Prediction to No-Regret Online Learning"
- Ho & Ermon (2016) - "Generative Adversarial Imitation Learning"

**RL from Demonstrations:**
- Hester et al. (2018) - "Deep Q-learning from Demonstrations"
- Nair et al. (2018) - "Overcoming Exploration in RL with Demonstrations"

---

## Related Issues

- #48 - Curiosity (use after BC to explore beyond experts)
- #49 - Curriculum (BC as first phase of curriculum)
- #50 - Entropy scheduling (less needed with BC init)
- #51 - Hierarchical (BC can pre-train sub-policies)
- #53 - Reduced action space (simpler for BC)
